rest_service.go

```diff
@@ -394,3 +443,239 @@ func (s *Service) closeManager(rw http.ResponseWriter, r *http.Request) {
+			StatusCode: http.StatusInternalServerError,
+			Message:    errors.Wrap(err, "problem in download setup").Error(),
+		})
+		return
+	}
+
+	gimlet.WriteJSON(rw, struct{}{})
+}
+
+// SetupDownloadMongoDBReleases performs necessary setup to download MongoDB with the given options.
+func SetupDownloadMongoDBReleases(ctx context.Context, cache *lru.Cache, opts MongoDBDownloadOptions) error {
+	if err := opts.BuildOpts.Validate(); err != nil {
+		return errors.Wrap(err, "invalid build options")
+	}
+
+	if err := MakeEnclosingDirectories(opts.Path); err != nil {
+		return errors.Wrap(err, "problem creating enclosing directories")
+	}
+
+	absRootPath, err := filepath.Abs(opts.Path)
+	if err != nil {
+		return errors.Errorf("problem getting absolute path to %s: %s", opts.Path, err)
+	}
+	addToCache := addMongoDBFilesToCache(cache, absRootPath)
+
+	feed, err := bond.GetArtifactsFeed(opts.Path)
+	if err != nil {
+		return errors.Wrap(err, "problem making artifacts feed")
+	}
+
+	catcher := grip.NewBasicCatcher()
+	urls, errs1 := feed.GetArchives(opts.Releases, opts.BuildOpts)
+	jobs, errs2 := createDownloadJobs(opts.Path, urls)
+
+	if err := setupDownloadJobsAsync(ctx, jobs, processDownloadJobs(context.Background(), addToCache)); err != nil {
+		catcher.Add(errors.Wrap(err, "problem starting download jobs"))
+	}
+
+	if err := aggregateErrors(errs1, errs2); err != nil {
+		catcher.Add(errors.Wrap(err, "problem initializing download jobs"))
+	}
+
+	return catcher.Resolve()
+}
+
+func setupDownloadJobsAsync(ctx context.Context, jobs <-chan amboy.Job, processJobs func(amboy.Queue) error) error {
+	q := queue.NewLocalUnordered(runtime.NumCPU())
+	if err := q.Start(ctx); err != nil {
+		return errors.Wrap(err, "problem starting download job queue")
+	}
+
+	if err := amboy.PopulateQueue(ctx, q, jobs); err != nil {
+		return errors.Wrap(err, "problem adding download jobs to queue")
+	}
+
+	go func() {
+		if err := processJobs(q); err != nil {
+			grip.Errorf(errors.Wrap(err, "error occurred while adding jobs to cache").Error())
+		}
+	}()
+
+	return nil
+}
+
+func addMongoDBFilesToCache(cache *lru.Cache, absRootPath string) func(string) error {
+	return func(fileName string) error {
+		catcher := grip.NewBasicCatcher()
+
+		filePath := filepath.Join(absRootPath, fileName)
+		if err := cache.AddFile(filePath); err != nil {
+			catcher.Add(errors.Wrap(err, "problem adding file to cache"))
+		}
+
+		baseName := filepath.Base(fileName)
+		ext := filepath.Ext(baseName)
+		dirPath := filepath.Join(absRootPath, string(baseName[:len(baseName)-len(ext)]))
+
+		err := filepath.Walk(dirPath, func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+
+			// Cache only handles individual files, not directories.
+			if !info.IsDir() {
+				if err := cache.AddStat(path, info); err != nil {
+					catcher.Add(errors.Wrap(err, "problem adding file to cache"))
+				}
+			}
+
+			return nil
+		})
+		if err != nil {
```

Comment:
This check is unnecessary. `errors.Wrap(nil)` returns nil and `catcher.Add(nil)` is a noop. You can just `catcher.Add(errors.Wrap(err, "problem walking download files"))` without the `if`.